### ISSUE TITLE
DNS plugin in 5.5.0 segfaults under Linux 4.1.x

### ISSUE THREAD

**Issue by hnyman** [collectd/collectd#1227]
I have built identical builds for Openwrt router (ar71xx / mips32r2 / WNDR3700) with Linux kernel 3.8.x and 4.1.x, and collectd 5.5.0 segfaults quickly with Linux 4.1.x if the dns plugin is enabled. If DNS plugin is disabled, collectd works otherwise normally also with 4.1.x. The plugin works ok with Linux 3.8.x.

Segfault:

```
[  104.435977] 
[  104.435977] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 76db3ae8
[  104.444585] epc = 76f6ef30 in dns.so[76f6e000+14000]
[  104.449597] ra  = 0040f084 in collectd[400000+30000]
[  104.454602] 
```

I have opened an issue at the Openwrt bug tracker ( https://github.com/openwrt/packages/issues/1660 ), but I am opening also this issue here as this sounds more like an collectd problem due to some change at Linux.

**Comment by rubenk** [comment]
Would it be possible for you to get a backtrace?
https://collectd.org/wiki/index.php/Core_file

**Comment by hnyman** [comment]
Looks difficult. I tried compiling with debug support and used gdb remote debugging to fetch info.
It doesn't seem to find any usable info for stack trace.
```
(gdb) c
Continuing.
[New Thread 3217.3231]
warning: GDB can't find the start of the function at 0x77ce4f38.

    GDB is unable to find the start of the function at 0x77ce4f38
and thus can't determine the size of that function's stack frame.
This means that GDB may be unable to access that stack frame, or
the frames below it.
    This problem is most likely caused by an invalid program counter or
stack pointer.
    However, if you think GDB should simply search farther back
from 0x77ce4f38 for code which looks like the beginning of a
function, you can increase the range of the search using the `set
heuristic-fence-post' command.

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 3217.3231]
0x77ce4f38 in ?? ()
(gdb) bt
#0  0x77ce4f38 in ?? ()
(gdb) c
Continuing.

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
(gdb) 


********************************
On the router the output (from collectd -f) stops at the start of the dns functionality:

pe = ping_droprate; type_instance = 82.181.192.1;
uc_insert: Added OpenWrt2/ping/ping_droprate-82.181.192.1 to the cache.
plugin: plugin_write: Writing values via rrdtool.
rrdtool plugin: rrd_cache_insert: file = /tmp/rrd/OpenWrt2/ping/ping_droprate-82.181.192.1.rrd; values_num = 1; age = 0.000;
rrdtool plugin: queue thread: Wrote 1 value to /tmp/rrd/OpenWrt2/ping/ping_droprate-82.181.192.1.rrd
plugin_read_thread: Effective interval of the ping plugin is 30.000 seconds.
plugin_read_thread: Next read of the ping plugin at 1439750315.262.
plugin_read_thread: Handling `dns'.
```

Ps. collectd does not crash 100% of the time, but stays alive maybe 10% of the tries. Strange.

**Comment by hnyman** [comment]
Well, after adjusting ulimit I was able to get core files and this post-mortem output from gdb. It seems to be a reference to plugin.c:
```
perus@vb1504:/Openwrt/trunk$ ./build_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/gdb-7.8/gdb/gdb ./staging_dir/target-mips_34kc_musl-1.1.10/root-ar71xx/usr/sbin/collectd ./collectd.3272.11.1439751032.core
GNU gdb (GDB) 7.8
...
This GDB was configured as "--host=x86_64-linux-gnu --target=mips-openwrt-linux-musl".
...
Reading symbols from ./staging_dir/target-mips_34kc_musl-1.1.10/root-ar71xx/usr/sbin/collectd...done.
[New LWP 3286]
[New LWP 3284]
[New LWP 3278]
[New LWP 3283]
[New LWP 3277]
[New LWP 3280]
[New LWP 3272]
[New LWP 3281]
[New LWP 3279]
[New LWP 3282]
[New LWP 3285]
Core was generated by `/usr/sbin/collectd -f'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x77ce4f38 in ?? ()
(gdb) bt
#0  0x77ce4f38 in ?? ()
#1  0x0040f5dc in plugin_read_thread (args=<optimized out>) at plugin.c:566
Backtrace stopped: Cannot access memory at address 0x77913b2c
(gdb) 
```
I am not very familiar with gdb, so any hints for further debugging?

**Comment by mfournier** [comment]
This pattern of error has been related to threading issues a few times in the past.

How many read plugins are you using and what is the value of the `ReadThreads` parameter ? Are you able to estimate how much time the dns plugin takes to read its values ? And what `Interval` do you have configured ?

Does the problem go away if you define `ReadThreads 1` ?

**Comment by hnyman** [comment]
Thanks for help.

This is a home router, so the read plugins are rather light-weight system status plugins:
10 plugins in total: entropy, conntrack, DNS, interface, ping, memory, uptime, cpu, load + Openwrt-specific wireless "iwinfo"

Interval is 30 seconds.
Readthreads is 2.

* Disabling other plugins, and trying only the DNS plugin with Readthreads 2 makes collectd to crash.
* But decreasing Readthreads to 1 enables DNS plugin to work. Both as the single read plugin and also with all 10 plugins enabled.

Apparently there is something in the DNS plugin which is not quite thread-safe (?) and which has now surfaced in my environment.

Do you see any performance issues running with Readthreads 1 ?

Should I try any further debugging?

Ps. I have no estimate on how long the DNS plugin takes to read values. Any practical advice, how to measure that?

**Comment by mfournier** [comment]
Ok, so at least we have the confirmation the issue is located in the dns plugin, or in one of the libs it bases on.

The problem with `ReadThreads 1` is that if one read operation takes a long time, all the other read plugins will block, waiting on the single read thread to become available.

I'm puzzled by the fact the problem only occurs with kernel 4.1.x. Are you able to re-test this on a 3.8.x kernel with the stripped down setup you now have (`Readthreads 2` and only the dns plugin enabled) ? Maybe setting a lower `Interval` would help trigger the issue ?

Are you able to share the relevant parts of your configuration ? Thanks !

**Comment by hnyman** [comment]
I will try again with kernel 3.8.x. I have old firmware images from July that I can flash back to the router.

But I am sorry to say that my earlier conclusion seems a bit premature:
I have tested several times with minimal configuration of DNS only or DNS+entropy plugins.
And sadly collectd may crash also with "Readthreads 1" setting. With both 1 and 2 it crashes even the plain DNS plugin craches every 3-6th time immediately at collectd startup. 

Below is a system log excerpt with a debug log line just before each collectd restart. I restarted the process manually after seeing from the log, if collectd had crashed with page_fault in the previous attempt. I used the line
``` logger "koe10 - rt1 dns" ; /etc/init.d/collectd restart ; logread | tail``` 

Both with readthreads 1 and 2, the crashiness varies a lot.
I didn't let the process run longer, so the test does not really measure if the DNS plugin would crash later.

One additional piece of information: Openwrt currently uses musl-libc as the C library. (The crashiness started during musl 1.10, but now it is 1.11. Old kernel 3.8 used musl 1.10) 

```
T
Thu Sep  3 15:08:03 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:03 2015 kern.info kernel: [11303.269483]
Thu Sep  3 15:08:03 2015 kern.info kernel: [11303.269483] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 77a3bae8
Thu Sep  3 15:08:03 2015 kern.info kernel: [11303.278090] epc = 77b90f30 in dns.so[77b90000+14000]
Thu Sep  3 15:08:03 2015 kern.info kernel: [11303.283106] ra  = 0040f084 in collectd[400000+30000]
Thu Sep  3 15:08:03 2015 kern.info kernel: [11303.288110]
Thu Sep  3 15:08:04 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:05 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:06 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:07 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:08 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:08 2015 kern.info kernel: [11308.588820]
Thu Sep  3 15:08:08 2015 kern.info kernel: [11308.588820] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 77541ae8
Thu Sep  3 15:08:08 2015 kern.info kernel: [11308.597393] epc = 77696f30 in dns.so[77696000+14000]
Thu Sep  3 15:08:08 2015 kern.info kernel: [11308.602384] ra  = 0040f084 in collectd[400000+30000]
Thu Sep  3 15:08:08 2015 kern.info kernel: [11308.607382]
Thu Sep  3 15:08:09 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:10 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:11 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:12 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:13 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:13 2015 kern.info kernel: [11313.301371]
Thu Sep  3 15:08:13 2015 kern.info kernel: [11313.301371] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 77233ae8
Thu Sep  3 15:08:13 2015 kern.info kernel: [11313.309955] epc = 77388f30 in dns.so[77388000+14000]
Thu Sep  3 15:08:13 2015 kern.info kernel: [11313.314947] ra  = 0040f084 in collectd[400000+30000]
Thu Sep  3 15:08:13 2015 kern.info kernel: [11313.319968]
Thu Sep  3 15:08:14 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:15 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:18 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:19 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:21 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:23 2015 user.notice root: koe10 - rt1 dns
Thu Sep  3 15:08:23 2015 kern.info kernel: [11323.628877]
Thu Sep  3 15:08:23 2015 kern.info kernel: [11323.628877] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 773b3ae8
Thu Sep  3 15:08:23 2015 kern.info kernel: [11323.637468] epc = 77508f30 in dns.so[77508000+14000]
Thu Sep  3 15:08:23 2015 kern.info kernel: [11323.642457] ra  = 0040f084 in collectd[400000+30000]
Thu Sep  3 15:08:23 2015 kern.info kernel: [11323.647453]
Thu Sep  3 15:08:25 2015 user.notice root: koe10 - rt1 dns

SWITCH TO  Readthreads 2

Thu Sep  3 15:12:59 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:01 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:02 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:03 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:06 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:07 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:07 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:08 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:09 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:10 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:10 2015 kern.info kernel: [11610.825697]
Thu Sep  3 15:13:10 2015 kern.info kernel: [11610.825697] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 77b47ae8
Thu Sep  3 15:13:10 2015 kern.info kernel: [11610.834269] epc = 77c9cf30 in dns.so[77c9c000+14000]
Thu Sep  3 15:13:10 2015 kern.info kernel: [11610.839262] ra  = 0040f084 in collectd[400000+30000]
Thu Sep  3 15:13:10 2015 kern.info kernel: [11610.844393]
Thu Sep  3 15:13:12 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:13 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:13 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:13 2015 kern.info kernel: [11613.993330]
Thu Sep  3 15:13:13 2015 kern.info kernel: [11613.993330] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 76e2bae8
Thu Sep  3 15:13:13 2015 kern.info kernel: [11614.001882] epc = 76f80f30 in dns.so[76f80000+14000]
Thu Sep  3 15:13:13 2015 kern.info kernel: [11614.006902] ra  = 0040f084 in collectd[400000+30000]
Thu Sep  3 15:13:13 2015 kern.info kernel: [11614.011859]
Thu Sep  3 15:13:15 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:17 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:18 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:20 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:20 2015 kern.info kernel: [11620.234244]
Thu Sep  3 15:13:20 2015 kern.info kernel: [11620.234244] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 77b47ae8
Thu Sep  3 15:13:20 2015 kern.info kernel: [11620.242823] epc = 77c9cf30 in dns.so[77c9c000+14000]
Thu Sep  3 15:13:20 2015 kern.info kernel: [11620.247837] ra  = 0040f084 in collectd[400000+30000]
Thu Sep  3 15:13:20 2015 kern.info kernel: [11620.252843]
Thu Sep  3 15:13:21 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:22 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:23 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:25 2015 user.notice root: koe11 - rt2 dns
Thu Sep  3 15:13:27 2015 user.notice root: koe11 - rt2 dns
```

**Comment by hnyman** [comment]
I flashed kernel 3.8 based firmware including collectd 5.5.0 and I can't get it to crash.
I tried starting collectd 20-30 times with 9 plugins enabled with each of the four settings:
* readthreads 1, interval 30s
* readthreads 2, interval 30s
* readthreads 2, interval 10s
* readthreads 5, interval 10s

No startup crashes with any of the ~120 times.

Like I said in my original message on Openwrt side, when comparing the latest non-crashing and earliest crashing version, I do not see any material/significant/possible code differences at the Openwrt side except the kernel change. 

Some possible ideas about the error source:
* The DNS plugin seems to handle the data collecting loop timing by itself, while some other plugins use threading library wait routines (or something tlike hat). I am not really specialist enough to see if the DNS plugin behaves in an unorthodox way or if its behaviour differs from other plugins.
* Openwrt has been so far using libpcap 1.5.3 from December 2013 (which library the DNS plugin utilises). I just noticed that Openwrt has been today updated to use 1.7.4 (April 2015 version) from now on, so I will test with that libpcap version contains some fixes for newer kernels.
* Programs for both 3.8 and 4.1 kernels have been compiled with the musl C-library, which also provides the support for threading. As musl is rather new, it is one possible source for the error. The kernel change might have brought something up regarding kernel threading compatibility. Pure guess, but possible. 

I will try with the new libpcap, and will also enable the "enable-debug" flag in collectd. I will possibly also insert more debugging log statements into collectd DNS fucntion entrances source code to see where it crashes.

Just for reference, the complete /etc/collect.conf :  
```
BaseDir "/var/run/collectd"
Include "/etc/collectd/conf.d"
PIDFile "/var/run/collectd.pid"
PluginDir "/usr/lib/collectd"
TypesDB "/usr/share/collectd/types.db"
Interval 10
ReadThreads 2


LoadPlugin conntrack

LoadPlugin rrdtool
<Plugin rrdtool>
        DataDir "/tmp/rrd"
        RRARows 100
        RRASingle true
        RRATimespan 3600
        RRATimespan 86400
        RRATimespan 604800
        RRATimespan 2678400
        RRATimespan 31622400
</Plugin>

LoadPlugin iwinfo
<Plugin iwinfo>
        IgnoreSelected false
</Plugin>

LoadPlugin interface
<Plugin interface>
        IgnoreSelected false
        Interface "br-lan"
</Plugin>

LoadPlugin ping
<Plugin ping>
        TTL 127
        Interval 30
        Host "87.92.0.1"
</Plugin>

LoadPlugin dns
<Plugin dns>
        Interface "br-lan"
        IgnoreSource "127.0.0.1"
</Plugin>

LoadPlugin memory

LoadPlugin uptime

LoadPlugin cpu

LoadPlugin load

```

**Comment by hnyman** [comment]
I tried with kernel 4.1 - libpcap 1.7.4 - DNS plugin only - collectd debug messages enabled.

I launched collectd manually from console without forking, so that the debug messages are shown.

After several tries I finally noticed a systematic difference:
* **when it crashes at startup, the first ```plugin_read_thread: Handling `dns'``` message is shown on console before the "PCAP object created" message has been shown.**

When it starts ok:
```
...
Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: Creating PCAP object..
dns plugin: PCAP object created.
plugin_read_thread: Handling `dns'.
Created new plugin context.
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441294925.019.
plugin_read_thread: Handling `dns'.
...
```

When it crashes at startup, the first "Handling dns" comes before the "PCAP object created" message.
```
Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: Creating PCAP object..
plugin_read_thread: Handling `dns'.
Created new plugin context.
dns plugin: PCAP object created.
Segmentation fault
```

To me that sounds like the first value reading try is activated before the libraries have been completely initialised, or something like that.

**Comment by hnyman** [comment]
After the realization of the first value reading cycle being made before the plugin initialisation has been completed, I searched for clues and found #1197, which gave me an idea:

If the DNS plugin takes a few more milliseconds to initialise libpcap, then the first read cycle should be slightly postponed in order to reasonably guarantee the completion of the initialisation and the creation of the callback function.

The patch in #1197 does not guarantee any delay in case the cycle happens to be just exactly at the interval limit, so I modified the idea and crafted my own patch to delay the first plugin read cycle by 5 seconds:

```
--- a/src/daemon/plugin.c
+++ b/src/daemon/plugin.c
@@ -1149,7 +1149,7 @@ static int plugin_insert_read (read_func
 	int status;
 	llentry_t *le;
 
-	rf->rf_next_read = cdtime ();
+	rf->rf_next_read = cdtime () + 5 * (cdtime_t) 1073741824;  // delay first read 5 s
 	rf->rf_effective_interval = rf->rf_interval;
 
 	pthread_mutex_lock (&read_lock);
```
**After applying the patch, I have not been able to make collectd to crash.** 

Still no idea, why kernel bump from 3.8. to 4.1 triggered the problem for me, but apparently some timing slightly changed.

**Comment by hnyman** [comment]
@mfournier : 
Do you have any comment on my diagnosis and the patch?
I have not seen any crashes since I implemented my patch. So, I think that the patch seems to work. 5 seconds is probably overkill, but I wanted to be "sure".

Should this kind of delay for the first data read cycle be implemented in collectd?

Before I submit any patch to collectd on Openwrt, I would like to understand if something is likely happening here at upstream.

**Comment by mfournier** [comment]
Hey, @hnyman thanks so much for digging into this !

The fact you're seeing the log messages in a different order when the crash happens and the delay you artificially add to work around the bug makes me think of some shared resource being accessed by concurrent threads.

Some background which might help debug this further: there are 2 key functions at play here: `dns_init()` which is called once at startup time *while collectd is still single-threaded* and `dns_read()` which is called once per interval from within one of the read-threads.

What strikes me by reading `src/dns.c` is that a worker thread which has nothing to do with the read threads gets created in `dns_init()`. This is the thread in which the "PCAP object" gets created. So my guess is that with 4.1 something is slowing down the creation of this object, meanwhile the rest of collectd's initialisation finishes, collectd spawns the read threads, and the one in charge of the dns plugin fires off before the "PCAP object" gets properly created.

As you seem to now have build collectd with `--enable-debug`, are you by any chance able to obtain another backtrace from gdb ?

That said, the fact upgrading from 3.8 to 4.1 make the pcap functions slower might be worth digging into too. It looks like a performance regression to me (which doesn't mean we shouldn't fix this issue in collectd!)

**Comment by mfournier** [comment]
Really not sure about this, but would you mind giving https://github.com/mfournier/collectd/commit/b7d8c1fe331a867660c48a397067777f6a3429a5 a try ? Thanks !

**Comment by hnyman** [comment]
I will try that tomorrow. At the first glance the idea of moving pcap object creation to init looks better to me.

I was already writing an answer to you while your patch idea arrived, so below are some earlier thoughts, slightly edited...

1)
I am not sure how to get more info out of gdb. The previous backtraces were already done with enable-debug activated. And the result seemed to be a too early call to a not-yet-defined callback function. I don't have enough gdb skills to get a better backtrace :-( 

2)
But I looked at the dns_init function and started to wonder if the "listen_thread_init" variable could be utilised in read_dns to check if the worker thread has completed the initialisation. If listen_thread_init = 1 would signal post-init status, then dns_read might start with a check for that. (EDIT: your patch might have changed the validity of this idea, but I leave it still here) 

3)
If dns_init is only called once while single-threaded, then it seems to me that the value of "listen_thread_init" is currently not really utilised anywhere. If I get it right, its value is zeroed after an error in dns_child_loop, and is set to 1 after init completion in dns_init, but the value is only checked earlier in dns_init, right? If that function is only called once and single-threaded, the whole exercise seems unnecessary. Might be a leftover from some earlier logic.  (EDIT: you added the pcap initialisation before checking the value of that variable. Intentionally?)

3b)
If dns_init is only called once (while still in single-threaded mode), why does the function start with a mutex? Just wondering if that is unnecessary...

**Comment by hnyman** [comment]
Your patch does not help. Collectd crashes, but a bit differently. 
And it also crashes with the 1 second delay added, so the patch breaks something.

I added one more debug line, and this is the output:
```
dns plugin: Creating PCAP object..  at 1548010818401169652
Created new plugin context.
dns plugin: PCAP object created.    at 1548010818408095759
dns plugin: Creating dns_child_loop at 1548010818408849207
Initialization complete, entering read-loop.
plugin_read_thread: Handling `dns'. at 1548010818410391559
Created new plugin context.
Created new plugin context.
Segmentation fault (core dumped)
```
I tried analysing the core dump with gdb, but backtrace led to the same line as previously. I don't reach the threads deeply enough to get anything more useful out.

**Comment by mfournier** [comment]
Yes, the patch was quick and dirty, I'm not surprised it has issues. But if you're unable to get the *previous* crash, it's already an improvement ;)

What are these `at 15480108184XXXXXXXX` numbers ? timestamps ? I'm curious to know how much time there is between each of these events. Are you still running with a 30s `Interval` ?

What I observe here is a `Next read of the dns plugin at ...` log message within one second after the `Handling `dns'` one, which misses in your case.

Could you try to find out why you aren't able to obtain a complete backtrace ? It would be really helpful. One thing you could try is to run collectd inside gdb:
```
gdb /usr/sbin/collectd
run -f
# wait for crash
bt
```

Also is there by any chance an easy way to reproduce the openwrt environment you're using in a VM or something ? Slightly unrelated: I'm asking because I wonder if it would be feasible to integrate this distro in our continuous integration system.

**Comment by hnyman** [comment]
I changed the timestamps to be the same as usually dispayed in collectd. This is the display then, with the original dns plugin without your patch.
```
Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: Creating PCAP object..  at 1441724417.497
dns plugin: PCAP object created.    at 1441724417.503
plugin_read_thread: Handling `dns'. at 1441724417.498
Created new plugin context.
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441724447.490.
plugin_read_thread: Handling `dns'. at 1441724447.490
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441724477.490.

Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: Creating PCAP object..  at 1441724472.057
dns plugin: PCAP object created.    at 1441724472.063
plugin_read_thread: Handling `dns'. at 1441724472.057
Created new plugin context.
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441724502.049.

--------------------------------------------------

Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: Creating PCAP object..  at 1441724499.202
plugin_read_thread: Handling `dns'. at 1441724499.202
Created new plugin context.
dns plugin: PCAP object created.    at 1441724499.230
Segmentation fault

Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: Creating PCAP object..  at 1441724578.682
plugin_read_thread: Handling `dns'. at 1441724578.683
Created new plugin context.
dns plugin: PCAP object created.    at 1441724578.710
Segmentation fault
```

Running gbd on the router would probably consume too much flash space etc. from the router. But I can run remote-gdb stub in the router, which enables me to get a bit further. After googling for more info on using gdb on multi-threads, I got the following:
```
Continuing.
[New Thread 4783.4793]
warning: GDB can't find the start of the function at 0x77daef78.

    GDB is unable to find the start of the function at 0x77daef78
and thus can't determine the size of that function's stack frame.
This means that GDB may be unable to access that stack frame, or
the frames below it.
    This problem is most likely caused by an invalid program counter or
stack pointer.
    However, if you think GDB should simply search farther back
from 0x77daef78 for code which looks like the beginning of a
function, you can increase the range of the search using the `set
heuristic-fence-post' command.

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 4783.4793]
0x77daef78 in ?? ()
(gdb) info threads
[New Thread 4783.4786]
[New Thread 4783.4787]
[New Thread 4783.4788]
[New Thread 4783.4789]
[New Thread 4783.4790]
[New Thread 4783.4791]
[New Thread 4783.4792]
[New Thread 4783.4794]
  Id   Target Id         Frame 
  10   Thread 4783.4794  __clone () at src/thread/mips/clone.s:19
  9    Thread 4783.4792  __syscall () at src/internal/mips/syscall.s:22
  8    Thread 4783.4791  __cp_end () at src/thread/mips/syscall_cp.s:36
  7    Thread 4783.4790  __cp_end () at src/thread/mips/syscall_cp.s:36
  6    Thread 4783.4789  __cp_end () at src/thread/mips/syscall_cp.s:36
  5    Thread 4783.4788  __cp_end () at src/thread/mips/syscall_cp.s:36
  4    Thread 4783.4787  __cp_end () at src/thread/mips/syscall_cp.s:36
  3    Thread 4783.4786  __cp_end () at src/thread/mips/syscall_cp.s:36
* 2    Thread 4783.4793  warning: GDB can't find the start of the function at 0x77daef78.
0x77daef78 in ?? ()
  1    Thread 4783.4783  __clone () at src/thread/mips/clone.s:19
warning: GDB can't find the start of the function at 0x77daef78.
(gdb) thread apply all bt 

Thread 10 (Thread 4783.4794):
#0  __clone () at src/thread/mips/clone.s:19
#1  0x77fc7770 in __pthread_create (res=0x0, attrp=<optimized out>, entry=0x0, 
    arg=0x77ff610c <__libc+28>) at src/thread/pthread_create.c:274
#2  0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC

Thread 9 (Thread 4783.4792):
#0  __syscall () at src/internal/mips/syscall.s:22
#1  0x77fa528c in __syscall6 (n=4181, f=0, warning: GDB can't find the start of the function at 0x77d78ba2.
e=28, d=<optimized out>, 
    c=<optimized out>, b=<optimized out>, a=<optimized out>)
    at ./arch/mips/syscall_arch.h:160
#2  setsockopt (fd=<optimized out>, level=<optimized out>, 
    optname=<optimized out>, optval=<optimized out>, optlen=28)
    at src/network/setsockopt.c:6
#3  0x77d78ba4 in ?? ()

Thread 8 (Thread 4783.4791):
#0  __cp_end () at src/thread/mips/syscall_cp.s:36
#1  0x77fc6198 in __syscall_cp_c (nr=4238, u=<optimized out>, v=128, w=2, x=0, 
    y=2, z=2010135624) at src/thread/pthread_cancel.c:42
#2  0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC

Thread 7 (Thread 4783.4790):
#0  __cp_end () at src/thread/mips/syscall_cp.s:36
#1  0x77fc6198 in __syscall_cp_c (nr=4238, u=<optimized out>, v=128, w=2, x=0, 
    y=2, z=2010225776) at src/thread/pthread_cancel.c:42
#2  0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC

Thread 6 (Thread 4783.4789):
#0  __cp_end () at src/thread/mips/syscall_cp.s:36
#1  0x77fc6198 in __syscall_cp_c (nr=4238, u=<optimized out>, v=128, w=2, x=0, 
    y=2, z=2010315888) at src/thread/pthread_cancel.c:42
#2  0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC

Thread 5 (Thread 4783.4788):
#0  __cp_end () at src/thread/mips/syscall_cp.s:36
#1  0x77fc6198 in __syscall_cp_c (nr=4238, u=<optimized out>, v=128, w=2, x=0, 
    y=2, z=2010406000) at src/thread/pthread_cancel.c:42
---Type <return> to continue, or q <return> to quit---
#2  0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC

Thread 4 (Thread 4783.4787):
#0  __cp_end () at src/thread/mips/syscall_cp.s:36
#1  0x77fc6198 in __syscall_cp_c (nr=4238, u=<optimized out>, v=128, w=2, x=0, 
    y=2, z=2010496112) at src/thread/pthread_cancel.c:42
#2  0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC

Thread 3 (Thread 4783.4786):
#0  __cp_end () at src/thread/mips/syscall_cp.s:36
#1  0x77fc6198 in __syscall_cp_c (nr=4238, u=<optimized out>, v=128, w=2, x=0, 
    y=2, z=2010586224) at src/thread/pthread_cancel.c:42
#2  0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC

Thread 2 (Thread 4783.4793):
warning: GDB can't find the start of the function at 0x77daef78.
#0  0x77daef78 in ?? ()

Thread 1 (Thread 4783.4783):
#0  __clone () at src/thread/mips/clone.s:19
#1  0x77fc7770 in __pthread_create (res=0x77f15184, attrp=<optimized out>, 
    entry=0x40f328 <plugin_read_thread>, arg=0x0)
    at src/thread/pthread_create.c:274
#2  0x0040e3ec in start_read_threads (num=2) at plugin.c:649
#3  plugin_init_all () at plugin.c:1661
#4  0x00404ccc in do_init () at collectd.c:328
#5  main (argc=<optimized out>, argv=<optimized out>) at collectd.c:694
warning: GDB can't find the start of the function at 0x77daef78.
(gdb) 
```

**Comment by hnyman** [comment]
Well, I added a few more debug print lines to dns.c and plugin.c, and it looks like the completion of the pcap_open_live function is the critical one. Its running time varies from 5ms to 26ms. If it takes the "longer time", collectd crashes as the first reading cycle apparently runs before the pcap object is ready. That function pcap_open_live comes from the pcap library, so I am not sure it anything can be done at the collectd side for that time variation itself.

But it needs to be somehow ensured that the first data reading cycle does not start too early.

Below are 3 ok runs and 2 crashes showing the timestamps. Again the original 5.5.0 code without your patch.
(the last ok run is interesting, because the first reading cycle starts before the pcap_open_live has completed, but as the completion comes only 6ms later, it is apparently still ok...)
```
OK

Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: before pcap_open_live   at 1441729802.528
dns plugin: after pcap_open_live    at 1441729802.533
plugin_read_thread: Handling `dns'. at 1441729802.528 rf_type 0
Created new plugin context.
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441729832.520.
dns plugin: after pcap_compile      at 1441729802.535
dns plugin: PCAP object created.    at 1441729802.535
dns plugin: dnstop_set_callback ok  at 1441729802.536
plugin_read_thread: Handling `dns'. at 1441729832.520 rf_type 0
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441729862.520.


Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: before pcap_open_live   at 1441729884.386
dns plugin: after pcap_open_live    at 1441729884.391
plugin_read_thread: Handling `dns'. at 1441729884.387 rf_type 0
Created new plugin context.
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441729914.379.
dns plugin: after pcap_compile      at 1441729884.393
dns plugin: PCAP object created.    at 1441729884.394
dns plugin: dnstop_set_callback ok  at 1441729884.395


Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: before pcap_open_live   at 1441730120.156
plugin_read_thread: Handling `dns'. at 1441730120.156 rf_type 0
Created new plugin context.
plugin_read_thread: Effective interval of the dns plugin is 30.000 seconds.
plugin_read_thread: Next read of the dns plugin at 1441730150.148.
dns plugin: after pcap_open_live    at 1441730120.162
dns plugin: after pcap_compile      at 1441730120.165
dns plugin: PCAP object created.    at 1441730120.165
dns plugin: dnstop_set_callback ok  at 1441730120.166

------

CRASHES

Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: before pcap_open_live   at 1441729893.553
plugin_read_thread: Handling `dns'. at 1441729893.553 rf_type 0
Created new plugin context.
dns plugin: after pcap_open_live    at 1441729893.580
dns plugin: after pcap_compile      at 1441729893.581
dns plugin: PCAP object created.    at 1441729893.582
dns plugin: dnstop_set_callback ok  at 1441729893.582
Segmentation fault (core dumped)


Initialization complete, entering read-loop.
Created new plugin context.
Created new plugin context.
dns plugin: before pcap_open_live   at 1441730133.044
plugin_read_thread: Handling `dns'. at 1441730133.044 rf_type 0
Created new plugin context.
dns plugin: after pcap_open_live    at 1441730133.071
dns plugin: after pcap_compile      at 1441730133.072
dns plugin: PCAP object created.    at 1441730133.073
dns plugin: dnstop_set_callback ok  at 1441730133.073
Segmentation fault (core dumped)
```

**Comment by mfournier** [comment]
I've tried artificially delaying the pcap object creation by adding a `sleep(2)` statement just before the `pcap_open_live()` call, and with or without my patch I'm unable to reproduce the problem. The read thread for the dns plugin simply waits on the pcap initialisation to finish.

So I'm wondering if you'd might try the same thing on the 3.8.x kernel you have ? This would at least tell us if there's a performace regression in 4.1.x which leads some sort of race condition, or if something really strange is going on with libpcap/libpthread when using 4.1.x (possibly on mips only?).

**Comment by hnyman** [comment]
I tried assigning more debug print statements, but got no special results. It looked like the address of the callback function dns_read was ok in plugin.c when the callback is made etc.

Like said earlier, the variation is in libpcap initialisation. This may well be something specific to libpcap in mips, using musl-libc (including threading functionality) and kernel tweak x, y and z, etc. If it is something like that, debugging would be difficult.

Building the old 3.8 based version with the debug functionality would require a bit too much effort, as I should time-warp the whole development toolchain two months back (and also time-warp all my own changes during that time.)

However, the DNS plugin in so non-essential to the core statistics that I have no interest to go more deeply into debugging this. The one-second delay before the first read cycle seems to cure things, and should be harmless otherwise.

If you implement #1197, that will also cure things (unless the process is launched exactly at the interval break).

I will probably mail that one-second-delay patch to the Openwrt development list, so that the collectd version there will get that patch.

Below is the patch for the 5.5 branch:
```
--- a/src/daemon/plugin.c
+++ b/src/daemon/plugin.c
@@ -1149,7 +1149,7 @@ static int plugin_insert_read (read_func
 	int status;
 	llentry_t *le;
 
-	rf->rf_next_read = cdtime ();
+	rf->rf_next_read = cdtime () + (cdtime_t) 1073741824; //delay first read 1s
 	rf->rf_effective_interval = rf->rf_interval;
 
 	pthread_mutex_lock (&read_lock);
```

**Comment by mfournier** [comment]
@hnyman I agree, debugging this is not easy at all :-(

Would you mind making the openwrt folks aware of our problem with the 4.1.x kernel ? As mentioned, I'm actually wondering if there isn't something broken at a lower level. Maybe somebody with more clue than ourselves can make sense of all this ?

It would just be a shame if the openwrt distro shipped with a workaround just for collectd, if there actually is a deeper problem which might affect other programs.

Thanks for all the time and energy you spent on this !

**Comment by rubenk** [comment]
@hnyman is still an issue with recent versions of collectd / the kernel?

**Comment by hnyman** [comment]
> is still an issue with recent versions of collectd / the kernel?

Sorry, I haven't evaluated that since 2015, as collectd in Openwrt now contains my patch (referenced above) that manually delays the first read cycle by 1 second. That hack was enough for keep things running ok.

**Comment by rubenk** [comment]
Thanks for the feedback @hnyman.
